File: source/Model/Node.java
/**
 * @ Author: Mo David
 * @ Create Time: 2024-07-16 16:33:56
 * @ Modified time: 2024-07-16 16:59:48
 * @ Description:
 * 
 * Nodes will always have a string id associated with them.
 * They also store a collection of references to other nodes via their ids.
 */

package Model;

import java.util.HashMap;

public class Node<Data> {
  
  // Data of the node
  private String id;
  private Data data;

  // References to adjacent nodes
  private HashMap<String, Node> prevNodes;
  private HashMap<String, Node> nextNodes;

  /**
   * Instantiates an instance of the Node class.
   * 
   * @param   { String }  The id associated with the node.
   * @param   { Data }    Any object of the given type, to be stored by the node.
   */
  public Node(String id, Data data) {

    // Save the node details
    this.id = id;
    this.data = data;

    // Init the data structs
    this.prevNodes = new HashMap<>();
    this.nextNodes = new HashMap<>();
  }

  /**
   * Retrieves the id of the given node.
   * 
   * @return  { String }  The id of the node.
   */
  public String getId() {
    return this.id;
  }

  /**
   * Returns a string array containing all the ids of the next nodes.
   * 
   * @return  { String[]  }   A string array with all next node ids. 
   */
  public String[] getNextNodes() {
    return (String[]) this.nextNodes.keySet().toArray();
  }

  /**
   * Returns a string array containing all the ids of the prev nodes.
   * 
   * @return  { String[]  }   A string array with all prev node ids. 
   */
  public String[] getPrevNodes() {
    return (String[]) this.prevNodes.keySet().toArray();
  }

  /**
   * Adds a node in the next map of the current node.
   * 
   * @param   { Node }  next  The node to append.
   */
  public void addNextNode(Node next) {
    String id = next.getId();

    this.nextNodes.put(id, next);
  }

  /**
   * Adds a node in the prev map of the current node.
   * 
   * @param   { Node }  prev  The node to append.
   */
  public void addPrevNode(Node prev) {
    String id = prev.getId();

    this.prevNodes.put(id, prev);
  }

  /**
   * Returns whether or not the node is in the next map of the current node.
   * 
   * @param   { Node }  next  The node to check for.
   */
  public boolean hasNextNode(Node next) {
    String id = next.getId();

    return this.nextNodes.containsKey(id);
  }

  /**
   * Returns whether or not the node is in the prev map of the current node.
   * 
   * @param   { Node }  prev  The node to check for.
   */
  public boolean hasPrevNode(Node prev) {
    String id = prev.getId();

    return this.prevNodes.containsKey(id);
  }
}

File: source/model/node.c
/**
 * @ Author: Mo David
 * @ Create Time: 2024-07-17 10:27:36
 * @ Modified time: 2024-07-24 22:16:33
 * @ Description:
 * 
 * The node class.
 */

#ifndef NODE_C
#define NODE_C

#include "./structs/hashmap.c"

#define NODE_ID_LENGTH (1 << 6)

typedef struct Node Node;

/**
 * Represents a node within our graph.
 * Uses a hashmap to store adjacencies.
 */
struct Node {

  // We use these variables for pathfinding
  Node *pPrevNode;
  Node *pNextNode;

  // We use this instead when creating undirected graphs
  HashMap *adjNodes;

  // The id and data of the node
  char id[NODE_ID_LENGTH + 1];
  void *pData;
};

/**
 * Allocates memory for a node.
 * 
 * @return  { Node * }  The memory for the new node.
 */
Node *_Node_alloc() {
  Node *pNode = calloc(1, sizeof(*pNode));
  
  return pNode; 
}

/**
 * Initializes the given node.
 * 
 * @param   { Node * }  this    The node to initialize.
 * @param   { char * }  id      The id of the node.
 * @param   { void * }  pData   The data stored by the node.
 * @return  { Node * }          The initialized node.
 */
Node *_Node_init(Node *this, char *id, void *pData) {

  // Save the id and data
  strncpy(this->id, id, NODE_ID_LENGTH);
  this->pData = pData;

  // Set the pointer to null
  this->pPrevNode = NULL;
  this->pNextNode = NULL;

  // Init the hashmaps
  this->adjNodes = HashMap_new();

  return this;
}

/**
 * Creates a new initialized node.
 * 
 * @param   { char * }  id      The id stored by the node.
 * @param   { void * }  pData   The data stored by the node.
 */
Node *Node_new(char *id, void *pData) {
  return _Node_init(_Node_alloc(), id, pData);
}

/**
 * Frees the memory associated with the node.
 * 
 * @param   { Node * }  this              The node to free.
 * @param   { int }     bShouldFreeData   Whether or not to free the data in the entries and node.
 */
void Node_kill(Node *this, int bShouldFreeData) {
  
  // Kill the hashmap BUT don't kill the actual nodes in it 
  // That's why we pass a 0 to the method
  HashMap_kill(this->adjNodes, 0);

  // Free the data associated with the node
  if(bShouldFreeData)
    free(this->pData);

  // Free the instance
  free(this);
}

/**
 * Sets the next node of the given node.
 * 
 * @param   { Node * }  this    The node to modify.
 * @param   { Node * }  pNext   A pointer to the next node.
*/
void Node_setNext(Node *this, Node *pNext) {
  this->pNextNode = pNext;
}

/**
 * Sets the prev node of the given node.
 * 
 * @param   { Node * }  this    The node to modify.
 * @param   { Node * }  pPrev   A pointer to the prev node.
*/
void Node_setPrev(Node *this, Node *pPrev) {
  this->pPrevNode = pPrev;
}

/**
 * Adds an adjacent node to the given node.
 * This method is used exclusively when addNext() and addPrev() are not.
 * In other words, we either use addNext() and addPrev() OR addAdj() based on whether we're constructing a un/directed graph.
 * 
 * @param   { Node * }  this  The first node in the adjacency.
 * @param   { Node * }  pAdj  The second node in the adjacency.
*/
void Node_addAdj(Node *this, Node *pAdj) {
  HashMap_put(this->adjNodes, pAdj->id, pAdj);
  HashMap_put(pAdj->adjNodes, this->id, this);
}

#endif